✨feat: readiness endpoint for the API project

wss-rbunning opened the pull request:
<!--
# Instructions

1. Pick a meaningful and result oriented title for your pull request. (Use sentence case, don't include any words indicating the way you achived this result, we just want the result. Keep it short < 70 characters)
  - Prefix the title with an emoji to categorize what is being done. (Copy-paste the emoji from the list below.)
  - Assign the appropriate label(s) to the pull request. (Use one of the labels from the list below.)
  
2. Enter a succinct description that says why the PR is necessary, and what it does.
  - Implement aspect X
  - Leave out feature Y because of A
  - Improve performance by B
  - Improve accessibility by C

3. Provide clear testing instructions that include any pertinent information, i.e. seat, roles, etc.
4. Include screenshots of your changes if they impact the UI (Before & After).
5. Update the preview link with your pull request number.
6. Include any JIRA tickets, design documents, GitHub issues, or other related items to the bottom of the PR.

# Available labels
UI
API
Evaluation Server
OLAP

# Emojis for categorizing pull requests

✨ New feature or functionality
🐛 Bugfix
🔥 P0 fix
✅ Tests
🚀 Performance improvements
🖍 CSS / Styling
📖 Documentation
🏗 Infrastructure / Tooling / Builds / CI
↩️ Reverting a previous change
🧹 Refactoring / Housekeeping
🔧 Package Upgrades / Maintenance
🗑️ Deleting code

-->

# Adds a readiness endpoint to the API project.
The endpoint will check:
- that MongoDB is available
- that Redis is available
- that Kafka is available (if running FeatBit pro)

*If any of these services are unavailable, it will return a 503.

## Why this is necessary

This endpoint will indicate that the API's dependencies are available and is is ready to process requests, not just that the API application is running. This is useful for having a Kubernetes readiness probe (https://spacelift.io/blog/kubernetes-readiness-probe).

## How to test

1. Start the docker containers for FeatBit (redis, Mongodb, kafka)
2. Start the API project
3. Curl http://localhost:5000/health/readiness. You should get a 200 status code.

### How to test MongoDB being down

1. Stop the MongoDB container
3. Curl http://localhost:5000/health/readiness. You should get a 503 status code.

### How to test Redis being down

1. Stop the Redis container
3. Curl http://localhost:5000/health/readiness. You should get a 503 status code.

### How to test Kafka being down

1. Set `IS_PRO` to `true`
2. Restart the API project
3. Stop the Kafka container
4. Curl http://localhost:5000/health/readiness. You should get a 503 status code.

deleteLater commented:
Hi @wss-rbunning thanks for your PR ❤️, we'll review it as soon as we have time.

deleteLater commented on modules/back-end/src/Infrastructure/Redis/RedisReadinessCheck.cs:
```diff
@@ -0,0 +1,27 @@
+    {
+        try
+        {
+            await _redisClient.GetDatabase().PingAsync();
```
We can add a method `Task<TimeSpan> PingAsync();`  to `IRedisClient` interface, then in the `DefaultRedisClient`

```csharp
public Task<TimeSpan> PingAsync() => GetDatabase().PingAsync();
```

For mock

```csharp
_mockedClient.Setup(client => client.PingAsync()).Returns(Task.FromResult(TimeSpan.Zero));

// or
_mockedClient.Setup(client => client.PingAsync()).ThrowsAsync(testRedisError);
```

wss-rbunning replied:
I have added `PingAsync` to the redis client and interface 👍

deleteLater reviewed (commented):
Hi @wss-rbunning  Sorry for the late reply, I've left some comments for you.

deleteLater commented on modules/back-end/src/Api/Setup/MiddlewaresRegister.cs:
```diff
@@ -17,6 +17,7 @@ public static WebApplication SetupMiddleware(this WebApplication app)
         // health check endpoints
         // external use
         app.MapHealthChecks("health/liveness", new HealthCheckOptions { Predicate = _ => false });
+        app.MapHealthChecks("health/readiness");
```
And when we add health checkers, we need to specify the `tags` for them

```csharp
// specify 'readniness' tag
builder.AddCheck<ReadinessCheck>("Check If Kafka Is Available", tags: new[] { "readiness" });
```

wss-rbunning replied:
I have added the tags to the checks and predicate for the readiness endpoint 👍

wss-rbunning commented on modules/back-end/src/Infrastructure/MongoDb/MongoDbClient.cs:
```diff
@@ -82,7 +82,7 @@ public MongoDbClient(IOptions<MongoDbOptions> options)
     };
 
-    public IMongoCollection<TEntity> CollectionOf<TEntity>()
+    public virtual IMongoCollection<TEntity> CollectionOf<TEntity>()
```
I have added `PingAsync` to the `MongoDbClient` 👍

deleteLater commented:
There is a build error.

```logs
/source/tests/Application.IntegrationTests/Stubs/TestRedisClient.cs(7,32): error CS0535: 'TestRedisClient' does not implement interface member 'IRedisClient.PingAsync()' [/source/tests/Application.IntegrationTests/Application.IntegrationTests.csproj]
```

deleteLater commented on modules/back-end/src/Infrastructure/Readiness/ReadinessExtensions.cs:
```diff
@@ -0,0 +1,29 @@
+    {
+        var readinessTags = new string[] { ReadinessTag };
+
+        builder.AddCheck<MongoDbReadinessCheck>("Check If MongoDB Is Available", tags: readinessTags)
```
We also need to specify a timeout for each readiness check, for example

```suggestion
        builder
              .AddCheck<MongoDbReadinessCheck>(
                  "Check If MongoDB Is Available",
                  tags: readinessTags,
                  timeout: TimeSpan.FromSeconds(5)
        )
```

deleteLater replied:
Without this, when I test MongoDB being down, the `/health/readniness`  endpoint spends **30 seconds** to return a `503` status. And if the redis being down, it spends nealy **6 seconds** to return a `503` status.

wss-rbunning replied:
Timeouts have been applied and set to 5 seconds 👍

deleteLater commented:
Another thing I want to mention is that actually we can use these nuget packages to do the health checks without changing much to our existing code.

- AspNetCore.HealthChecks.MongoDb
- AspNetCore.HealthChecks.Redis
- AspNetCore.HealthChecks.Kafka

From https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks

wss-rbunning commented:
> Another thing I forgot to mention is that actually we can use these nuget packages to do the health checks without changing much to our existing code.
> 
> * AspNetCore.HealthChecks.MongoDb
> * AspNetCore.HealthChecks.Redis
> * AspNetCore.HealthChecks.Kafka
> 
> From https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks

This is cool. I wasn't aware of these NuGet packages. I have switched over to using them 👍 .

Unfortunately, the Kafka one doesn't work with separate hosts set for the consumer and producer. It currently only supports taking a ProducerConfig, but even if you try to register another check to check the consumer, it will only check one of them. If the producer and consumer are set to separate hosts and one of them is down but not the other, FeatBit should be considered not ready. I have left my check for the consumer in and deleted any unused code 👍 .

AspNetCore.HealthChecks.Kafka will have to be updated to support that configuration.

deleteLater commented:
I noticed some logs and fixed them in https://github.com/featbit/featbit/pull/645/commits/645515bdb54fe1841939c03ae27a1b51d0e7cdc7

```log
%4|1718161273.027|CONFWARN|rdkafka#producer-2| [thrd:app]: Configuration property group.id is a consumer property and will be ignored by this producer instance
%4|1718161273.027|CONFWARN|rdkafka#producer-2| [thrd:app]: Configuration property enable.auto.commit is a consumer property and will be ignored by this producer instance
%4|1718161273.027|CONFWARN|rdkafka#producer-2| [thrd:app]: Configuration property auto.commit.interval.ms is a consumer property and will be ignored by this producer instance
%4|1718161273.027|CONFWARN|rdkafka#producer-2| [thrd:app]: Configuration property enable.auto.offset.store is a consumer property and will be ignored by this producer instance
%4|1718161273.027|CONFWARN|rdkafka#producer-2| [thrd:app]: Configuration property auto.offset.reset is a consumer property and will be ignored by this producer instance
```

deleteLater commented:
As for the kafka, I tried this and it works as expected. 

```csharp
if (configuration.IsProVersion())
{
    var producerHost = configuration.GetValue<string>("Kafka:Producer:bootstrap.servers");
    var consumerHost = configuration.GetValue<string>("Kafka:Consumer:bootstrap.servers");

    builder
        .AddKafka(
            new ProducerConfig
            {
                BootstrapServers = consumerHost
            },
            name: "Check if Kafka consumer is available.",
            tags: readinessTags,
            timeout: timeoutFiveSeconds
        )
        .AddKafka(
            new ProducerConfig
            {
                BootstrapServers = producerHost
            },
            name: "Check if Kafka producer is available.",
            tags: readinessTags,
            timeout: timeoutFiveSeconds
        );
}
```

deleteLater commented:
Here is how I test this PR

## Setup infrastructures

**For standard version**

```
docker compose up -d mongodb redis
```

**For pro version**

1. Add two extra services to `docker-compose-pro.yml` file

```yml
...
kafka-producer:
    image: bitnami/kafka:3.5
    restart: on-failure
    ports:
      - '9192:9192'
      - '29192:29192'
    environment:
      KAFKA_CFG_NODE_ID: 0
      KAFKA_CFG_PROCESS_ROLES: controller,broker
      KAFKA_CFG_CONTROLLER_QUORUM_VOTERS: 0@kafka-producer:9193
      KAFKA_CFG_LISTENER_SECURITY_PROTOCOL_MAP: PLAINTEXT:PLAINTEXT,PLAINTEXT_HOST:PLAINTEXT,CONTROLLER:PLAINTEXT
      KAFKA_CFG_LISTENERS: PLAINTEXT://:9192,PLAINTEXT_HOST://:29192,CONTROLLER://:9193
      KAFKA_CFG_ADVERTISED_LISTENERS: PLAINTEXT://kafka-producer:9192,PLAINTEXT_HOST://localhost:29192
      KAFKA_CFG_CONTROLLER_LISTENER_NAMES: CONTROLLER
      KAFKA_CFG_AUTO_CREATE_TOPICS_ENABLE: 'true'
      ALLOW_PLAINTEXT_LISTENER: 'true'
    networks:
      - featbit-network
    volumes:
      - kafka:/bitnami/kafka-consumer

kafka-ui:
  image: provectuslabs/kafka-ui:latest
  container_name: kafka-ui
  ports:
    - "8080:8080"
  restart: on-failure
  depends_on:
    - kafka
  environment:
    - KAFKA_CLUSTERS_0_NAME=featbit-kafka
    - KAFKA_CLUSTERS_0_BOOTSTRAPSERVERS=kafka:9092
    - KAFKA_CLUSTERS_1_NAME=featbit-kafka-producer
    - KAFKA_CLUSTERS_1_BOOTSTRAPSERVERS=kafka-producer:9192
  networks:
    - featbit-network
...
```

2. Update `appsettings.Development.json` file

- set `IS_PRO` to `true`
- set `Kafka:Producer:bootstrap.servers` to `localhost:29192`

3. Start services
```bash
docker compose -f docker-compose-pro.yml up mongodb redis kafka kafka-producer kafka-ui -d
```

## Test

1. Start Api project
2. Check readiness with `curl -v -w '\n* Response time: %{time_total}s\n' http://localhost:5000/health/readiness`
3. Stop any infrastructure, perform the check, the response should be returned in around 5 seconds with a 503 status code

deleteLater commented:
@wss-rbrennan I've pushed some commits to your branch. If that works for your senario, then we can merge this PR.

wss-rbunning commented:
> > It currently only supports taking a ProducerConfig, but even if you try to register another check to check the consumer, it will only check one of them
> > AspNetCore.HealthChecks.Kafka will have to be updated to support that configuration.
> 
> I am not sure I understand what you said correctly, but I tried this and it works as expected. [306a12f](https://github.com/featbit/featbit/commit/306a12f6146cc0bfbfd0020b27a9b205d05e0d5f)
> 
> ```cs
> if (configuration.IsProVersion())
> {
>     var producerHost = configuration.GetValue<string>("Kafka:Producer:bootstrap.servers");
>     var consumerHost = configuration.GetValue<string>("Kafka:Consumer:bootstrap.servers");
> 
>     builder
>         .AddKafka(
>             new ProducerConfig
>             {
>                 BootstrapServers = consumerHost
>             },
>             name: "Check if Kafka consumer is available.",
>             tags: readinessTags,
>             timeout: timeoutFiveSeconds
>         )
>         .AddKafka(
>             new ProducerConfig
>             {
>                 BootstrapServers = producerHost
>             },
>             name: "Check if Kafka producer is available.",
>             tags: readinessTags,
>             timeout: timeoutFiveSeconds
>         );
> }
> ```
> 
> The `AspNetCore.HealthChecks.Kafka` package checks the health of a kafka cluster by sending a message to the cluster and checking if the message has been persisted.
> 
> https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/netcore-6.0/src/HealthChecks.Kafka/KafkaHealthCheck.cs

I think I made a mistake in my previous testing. I am seeing the results you are now 👍 

I think when testing I installed the 8.0.1 version of the package. That version actually has a bug in it. [They register the health check as a singleton](https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/2f6a10f89e3c6d97f232f4157a80e3a9e1470dc5/src/HealthChecks.Kafka/DependencyInjection/KafkaHealthCheckBuilderExtensions.cs#L38C9-L38C137), which means that multiple calls to `AddKafka` aren't going to work. We should be good for now, but if FeatBit ever updates the version of that package it targets, the consumer Kafka check will not work.

To replicate the issue:

- Update `AspNetCore.HealthChecks.Kafka` to 8.0.1 and `Confluent.Kafka` to 2.3.0
- Start the API server according to your testing instructions
- Stop the consumer kafa container and curl the readiness endpoint. It should say it is healthy, even though it is not.

wss-rbunning commented:
> @wss-rbrennan I've pushed some commits to your branch. If that works for your senario, then we can merge this PR.

We should be good to merge 👍

deleteLater commented:
> I think I made a mistake in my previous testing. I am seeing the results you are now 👍
> 
> I think when testing I installed the 8.0.1 version of the package. That version actually has a bug in it. [They register the health check as a singleton](https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/2f6a10f89e3c6d97f232f4157a80e3a9e1470dc5/src/HealthChecks.Kafka/DependencyInjection/KafkaHealthCheckBuilderExtensions.cs#L38C9-L38C137), which means that multiple calls to `AddKafka` aren't going to work. We should be good for now, but if FeatBit ever updates the version of that package it targets, the consumer Kafka check will not work.

Yeah, you're right. They introduce the BUG in https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/pull/1440 which actually is unnecessary. So for now we better write our own `AddKafka` extension methods.

deleteLater commented:
I'm going to merge this PR if https://github.com/featbit/featbit/pull/645/commits/b2ac72e9cc92754da4a828f6810f9c77ed1c4f2d looks good to you

wss-rbunning commented:
> I'm going to merge this PR if [b2ac72e](https://github.com/featbit/featbit/commit/b2ac72e9cc92754da4a828f6810f9c77ed1c4f2d) looks good to you

Everything looks good 👍

deleteLater reviewed (approved):
LGTM, Thanks again for your time on this PR ♥️
